conn/physic/units_test.go

```diff
@@ -536,3 +537,166 @@ func BenchmarkCelsiusFloatg(b *testing.B) {
 		buf.Reset()
 	}
 }
+
+func TestAtod(t *testing.T) {
+	const (
+		negative = true
+		postive  = false
+	)
+	tests := []struct {
+		s    string
+		want decimal
+		used int
+		err  bool
+	}{
+		{"123456789", decimal{"123456789", 0, postive}, 9, false},
+		{"1nM", decimal{"1", 0, postive}, 1, false},
+		{"2.2nM", decimal{"22", -1, postive}, 3, false},
+		{"12.5mA", decimal{"125", -1, postive}, 4, false},
+		{"-12.5mA", decimal{"125", -1, negative}, 5, false},
+		{"1.1.1", decimal{}, 0, true},
+		{"1ma1", decimal{"1", 0, postive}, 1, false},
+		{"-0.00001%rH", decimal{"1", -5, negative}, 8, false},
+		{"0.00001%rH", decimal{"1", -5, postive}, 7, false},
+		{"--1ma1", decimal{"1", 0, negative}, 3, false},
+		{"++100ma1", decimal{"1", 2, postive}, 5, false},
+		{"1.0", decimal{"1", 0, postive}, 3, false},
+		{"0.10001", decimal{"10001", -5, postive}, 7, false},
+		{"-0.10001", decimal{"10001", -5, negative}, 8, false},
+		{"%-0.10001", decimal{"10001", -5, negative}, 0, true},
+		{"1n", decimal{"1", 0, postive}, 1, false},
+		{"200n", decimal{"2", 2, postive}, 3, false},
+	}
+
+	for _, tt := range tests {
+		got, n, err := atod(tt.s)
+
+		if got != tt.want && !tt.err {
+			t.Errorf("got %v expected %v", got, tt.want)
+		}
+		if tt.err && err == nil {
+			t.Errorf("expected error %v but got nil", err)
+		}
+
+		if n != tt.used {
+			t.Errorf("expected to consume %d char but used %d", tt.used, n)
+		}
+	}
+}
+
+func TestDoti(t *testing.T) {
+	tests := []struct {
+		name string
```

Comment:
Similar,
```
name     string
in       decimal
expected int64
```

Reply:
also was not checking for large -ve exponent.
so moved mag check bellow.
now:
```golang
mag := d.exp + scale
if mag < 0 {
	mag *= -1
}
if mag > 18 {
	return 0, errors.New("exponent exceeds int64")
}
```